lmdeploy/pytorch/models/mllama.py

```diff
@@ -0,0 +1,758 @@
+                            attn_metadata: Any, input_ids: torch.LongTensor):
+        # since every state share the same shape
+        cross_attention_states = torch.cat(cross_attention_states, 0)
+        full_text_row_masked_out_mask = torch.ones(
+            (attn_metadata.q_seqlens.sum(), 1), dtype=torch.bool)
+        start_pos = 0
+        img_idx = torch.where(input_ids == MLLAMA_IMAGE_TOKEN_ID)[1]
+        for img_id, q_seq_len in zip(img_idx.cpu(),
+                                     attn_metadata.q_seqlens.cpu()):
+            full_text_row_masked_out_mask[start_pos:img_id] = False
+            start_pos += q_seq_len
+        full_text_row_masked_out_mask = full_text_row_masked_out_mask.to(
+            cross_attention_states.device)
+
+        return cross_attention_states, full_text_row_masked_out_mask
+
+    def forward(
+        self,
+        input_ids: torch.Tensor,
+        position_ids: torch.Tensor,
+        past_key_values: List[List[torch.Tensor]],
+        cross_attention_states: Optional[torch.Tensor] = None,
+        attn_metadata: Any = None,
+        inputs_embeds: torch.Tensor = None,
+        cross_attn_metadata: Any = None,
+        **kwargs,
+    ):
+        """model forward, return logits."""
+        if cross_attn_metadata is None:
+            full_text_row_masked_out_mask = None
+        # FIXME basically, we want to inference
+        # text requests and image requests separately
+        elif cross_attention_states is None and (
+                cross_attn_metadata.kv_seqlens is None
+                or int(cross_attn_metadata.kv_seqlens.sum()) == 0):
+            full_text_row_masked_out_mask = None
+        elif cross_attn_metadata.is_decoding:
+            cross_attention_states = None
+            full_text_row_masked_out_mask = torch.ones(
+                (attn_metadata.q_seqlens.sum(), 1),
+                dtype=torch.bool,
+                device=input_ids.device)
+        else:
+            cross_attention_states, full_text_row_masked_out_mask = \
+                self.flat_encoder_result(cross_attention_states, cross_attn_metadata, input_ids)  # noqa
+        hidden_states = self.language_model(
+            input_ids=input_ids,
+            position_ids=position_ids,
+            cross_attention_states=cross_attention_states,
+            full_text_row_masked_out_mask=full_text_row_masked_out_mask,
+            past_key_values=past_key_values,
+            attn_metadata=attn_metadata,
+            inputs_embeds=inputs_embeds,
+            cross_attn_metadata=cross_attn_metadata,
+        )
+        return hidden_states
+
+    def get_logits(self, hidden_states: torch.Tensor):
+        """compute logits of the model output."""
+        return self.language_model.get_logits(hidden_states)
+
+    def support_cuda_graph(
+        self,
+        input_ids: torch.Tensor,
+        **kwargs,
+    ):
+        """support cudagraph."""
```

Comment:
can we enable cudagraph in decoding stage?

Reply:
The CUDA graph may not be applicable since inputs with and without images result in different graphs. For instance, if the first round input is text, the constructed CUDA graph might proceed to branch A in the model. However, in the second round, if the inputs include an image, the graph needs to go to branch B.

Reply:
ok